### PR TITLE
Add RMM devcontainers

### PR DESCRIPTION
**Pull request by trxcllnt**
## Description
This PR adds some [devcontainers](https://containers.dev/) to help simplify building the RMM C++ and Python libraries.

It also adds an optional job to the `pr.yaml` to [build the RMM libs in each devcontainer](https://github.com/trxcllnt/rmm/blob/fea/devcontainers/.github/workflows/pr.yaml#L79-L85), so the build caches are populated for devs by CI.

A devcontainer can be launched by clicking the "Reopen in Container" button that VSCode shows when opening the repo (or by using the "Rebuild and Reopen in Container" command from the command palette):
image

Clicking this button will cause VSCode to prompt the user to select one of these devcontainer variants:
image

On startup, the devcontainer creates or updates the conda/pip environment using `rmm/dependencies.yaml`. The envs/package caches are cached on the host via volume mounts, which are described in more detail in [`.devcontainer/README.md`](https://github.com/trxcllnt/rmm/blob/fea/devcontainers/.devcontainer/README.md).

The container includes convenience functions to clean, configure, and build the various RMM components:

```shell
$ clean-rmm-cpp # only cleans the C++ build dir
$ clean-rmm-python # only cleans the Python build dir
$ clean-rmm # cleans both C++ and Python build dirs

$ configure-rmm-cpp # only configures rmm C++ lib

$ build-rmm-cpp # only builds rmm C++ lib
$ build-rmm-python # only builds rmm Python lib
$ build-rmm # builds both C++ and Python libs
```

* The C++ build script is a small wrapper around `cmake -S ~/rmm -B ~/rmm/build` and `cmake --build ~/rmm/build`
* The Python build script is a small wrapper around `pip install --editable ~/rmm`

Unlike `build.sh`, these convenience scripts *don't* install the libraries after building them. Instead, they automatically inject the correct arguments to build the C++ libraries from source and use their build dirs as package roots:

```shell
$ cmake -S ~/rmm -B ~/rmm/build
$ CMAKE_ARGS="-Drmm_ROOT=~/rmm/build" \ # <-- this argument is automatic
  pip install -e ~/rmm
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by harrism — APPROVED** [review]
I know I've asked you this before... :) there is a lot of duplication across the multiple devcontainer.json files. And multiply that times the number of RAPIDS repos. Would it be possible to create another devcontainer feature that can be used to factor out the duplicated parts?

**Comment by trxcllnt** [comment]
> Would it be possible to create another devcontainer feature that can be used to factor out the duplicated parts?

Yes and no.

Yes, because just about all the things that can be factored out have been. We could factor "mounts" into a feature that's symlinked + referenced by each of the devcontainers, but I assumed hiding that from users wouldn't buy much.

No, because either the field is specific to a repo or devcontainer (the `build.args` list, etc.), or due to various bugs. For example, Codespaces only runs the `postAttachCommand` defined in the final `devcontainer.json`, not ones defined as part of a feature.

File: .github/workflows/pr.yaml
Comment (trxcllnt) [reply .github/workflows/pr.yaml:81]: Yeah, I'm testing all the repos can build the devcontainers using the workflow at this branch, but once they're all building we should merge that workflow and update these PRs.

File: .github/workflows/pr.yaml
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:30]:
```suggestion
```

I'd like to avoid switching this to `false` since we typically try not to add optional jobs to our PR workflows.

The `devcontainer` job is an exception to this rule.

Can you add a new input called `optional_jobs` that accepts a multi-line string of jobs that should not be considered required?

usage example:

```yaml
uses: rapidsai/shared-action-workflows/.github/workflows/checks.yaml@branch-23.10
    with:
      enable_check_generated_files: false
      optional_jobs: |
        devcontainer
```

**Review by ajschmidt8 — COMMENTED** [review]
Leaving one comment for now until I have time later for a proper review.

File: ci/release/update-version.sh
Comment (ajschmidt8) [inline ci/release/update-version.sh:61]: A few comments here:

- Is the `.devcontainer/Dockerfile` `sed` command necessary? I don't see any RAPIDS versions in that Dockerfile. Not sure if this was refactored out at some point
- For the reasons outlined in [this comment](https://github.com/rapidsai/ucxx/pull/36#discussion_r1178429239), I've been discouraging the use of variables (e.g. `CURRENT_SHORT_TAG`) in the search pattern of these `sed` commands since it makes the commands fragile. Additionally, these `update-version.sh` scripts will likely be refactored in the future to be passed a single argument, `RAPIDS_VERSION`, instead of distinguishing between "current" and "next" versions.
- Historically, we've intentionally used `sed_runner` instead of `sed` since it works Mac and Linux machines. This is important because the `update-version.sh` currently gets run locally by the Ops team @ release time and therefore needs to work on both machine types. We have some plans to revamp how these updates are handled in the future, but for now, it's probably best to stick to the status quo to prevent any unforeseen issues.

**Review by ajschmidt8 — COMMENTED** [review]
few more comments. I think that's all from me.

File: ci/release/update-version.sh
Comment (trxcllnt) [reply ci/release/update-version.sh:61]: > Is the .devcontainer/Dockerfile sed command necessary?

It was at one point, but the RAPIDS version was moved into each devcontainer.json file.

> Additionally, these update-version.sh scripts will likely be refactored in the future to be passed a single argument, `RAPIDS_VERSION`, instead of distinguishing between "current" and "next" versions.

That's fine, I can update the script to match explicit strings.

> Historically, we've intentionally used sed_runner instead of sed since it works Mac and Linux machines.

Ok, I'll make the `find` exec line do what the `sed_runner` does. I think the only alternative would be exporting `sed_runner` to the current env, and that sounds less desirable.

File: dependencies.yaml
Comment (bdice) [inline dependencies.yaml:172]: Align with #1343.
```suggestion
          - clang==16.0.6
          - clang-tools==16.0.6
```

File: dependencies.yaml
Comment (bdice) [inline dependencies.yaml:128]: #1344 is a duplicate, we can close that when this merges (or merge that before this PR, either way)

**Review by bdice — APPROVED** [review]
LGTM, once we update to clang 16.0.6.

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]
requesting some small changes

**Review by ajschmidt8 — APPROVED** [review]

**Pull request merged**
